bin.src/generateCommands.py

```diff
@@ -0,0 +1,581 @@
+class ReduceExposureStep(ScienceStep):
+    commandName = "reduceExposure.py"
+
+@ScienceStep.registerAsSubclass("mergeArms")
+class MergeArmsStep(ScienceStep):
+    commandName = "mergeArms.py"
+
+@ScienceStep.registerAsSubclass("calculateReferenceFlux")
+class CalculateReferenceFluxStep(ScienceStep):
+    commandName = "calculateReferenceFlux.py"
+
+@ScienceStep.registerAsSubclass("fluxCalibrate")
+class FluxCalibrateStep(ScienceStep):
+    commandName = "fluxCalibrate.py"
+
+@ScienceStep.registerAsSubclass("coaddSpectra")
+class CoaddSpectraStep(ScienceStep):
+    commandName = "coaddSpectra.py"
+
+class ScienceBlock:
+    steps = [] # will be replaced in processYaml()
+
+    def __init__(self, blockName:str, yamlBlock:Mapping[str,Any]):
+        self.name = blockName
+        self.visits = _getVisits(yamlBlock)
+
+        self.pipeline = {}
+        for step in self.steps:
+            block = yamlBlock.get("pipeline", {}).get(step, {})
+            self.pipeline[step] = ScienceStep.getSubclass(step).fromYaml(block)
+
+    def execute(self, dataDir, calib, rerun, steps=None, processes=1, verbose=False):
+        """Execute a science block"""
+
+        if steps is None:
+            steps = self.steps
+
+        if verbose:
+            print(f"Processing science block '{self.name}'", file=sys.stderr)
+
+        for step in self.steps:
+            self.pipeline[step].execute(self.visits, dataDir, calib, f"{rerun}/pipeline", processes=processes, verbose=verbose)
+
+
+def _getConfigs(yamlBlock:Mapping[str,Any]) -> List[str]:
+    """
+    Extract config: ... and configs: ... from yamlBlock.
+    """
+    configs = []
+
+    if "config" in yamlBlock:
+        configs.append(yamlBlock["config"])
+
+    if "configs" in yamlBlock:
+        for c in yamlBlock["configs"]:
+            configs.append(c)
+
+    return configs
+
+def _getVisits(yamlBlock:Mapping[str,Any]) -> List[int]:
+    """
+    Extract visit: ... and visits: ... from yamlBlock.
+    """
+    visits = []
+
+    if "visit" in yamlBlock:
+        visits.append(int(yamlBlock["visit"]))
+
+    if "visits" in yamlBlock:
+        for v in yamlBlock["visits"]:
+            if isinstance(v, str):
+                mat = re.match(r"^(\d+)\.\.(\d+)(:\d+)?$", v)
+                if mat:
+                    v1, v2, stride = mat.groups()
+                    v1 = int(v1)
+                    v2 = int(v2)
+                    stride = 1 if stride is None else int(stride[1:])
+                else:
+                    v1 = int(v)
+                    v2 = v1
+                    stride = 1
+
+                for v in range(v1, v2 + 1, stride):
+                    visits.append(v)
+            else:
+                visits.append(v)
+
+    return visits
+
+
+def ingestCalibs(calibType, dataDir, calib, rerun, copyMode, *, overwrite=False):
+    command = [
+        "ingestCalibs.py",
+        dataDir,
+        f"--output={os.path.join(dataDir, calib)}",
+        "--validity=1800",
```

Comment:
We must be able to specify the validity range, and it should be allowed to be different for different calib types.

Reply:
Now `validity` can be written in the YAML file. If not specified, it defaults to 1800.